src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1870 @@
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //
+    // Big case default:  registers on entry
+    //
+    //  rbx: haystack
+    //  rcx: k
+    //  rdx: junk
+    //  rsi: n
+    //  rdi: haystack
+    //  r10: n - k
+    //  r12: k
+    //  r13: junk
+    //  r14: needle
+    //  rbp: -1
+    //  XMM_BYTE_0 - first element of needle broadcast
+    //  XMM_BYTE_K - last element of needle broadcast
+    //
+    // Set up in big_case_loop_helper
+    //  XMM_BYTE_1 - second element of needle broadcast
+
+    __ bind(L_bigCaseDefault);
+
+    // Loop construct handling for big haystacks
+    // The helper binds L_loopTop which should be jumped to if potential matches fail to compare
+    // equal (thus moving on to the next chunk of haystack).  If we run out of haystack, the
+    // helper jumps to L_checkRangeAndReturn with a (-1) return value.
+    big_case_loop_helper(false, 0, L_checkRangeAndReturn, L_loopTop, mask, hsPtrRet, needleLen,
+                         needle, haystack, hsLength, tmp1, tmp2, tmp3, rScratch, ae, _masm);
+
+    // big_case_loop_helper will fall through to this point if one or more potential matches are found
+    // The mask will have a bitmask indicating the position of the potential matches within the haystack
+    __ align(OptoLoopAlignment);
+    __ bind(L_innerLoop);
+    __ tzcntl(index, mask);
+
+// Re-use of r15 and rdx
+#undef tmp2
+#undef tmp3
+#define retval r15
+#define firstNeedleCompare rdx
+
+    // Note that we're comparing the full needle here even though in some paths
+    // the 1st, 2nd, and last bytes are already known to be equal.  This is necessary
+    // due to the handling of cases where nMinusK is < 32
+
+    // Need a lot of registers here to preserve state across arrays_equals call
+
+    // Starting address in the haystack
+    __ leaq(haystackStart, Address(hsPtrRet, index, Address::times_1, isU ? 4 : 2));
+        // Starting address of first byte of needle to compare
+    __ leaq(firstNeedleCompare, Address(needle, isU ? 4 : 2));
+        // Number of bytes to compare
+    __ leaq(compLen, Address(needleLen, isU ? -6 : -3));
+
+    // Call arrays_equals for both UU and LL cases as bytes should compare exact
+    __C2 arrays_equals(false, haystackStart, firstNeedleCompare, compLen, retval, rScratch,
+                        XMM_TMP3, XMM_TMP4, false /* char */, knoreg);
+    __ testl(retval, retval);
+    __ jne_b(L_found);
+
+    // If more potential matches, continue at inner loop, otherwise go get another vector
+    CLEAR_BIT(mask);
+    __ jne(L_innerLoop);
+    __ jmp(L_loopTop);
+
+    // Found exact match.  Compute offset from beginning of haystack
+    __ bind(L_found);
+    __ subq(hsPtrRet, haystack);
+    __ addq(hsPtrRet, index);
+    __ movq(r11, hsPtrRet);
+    __ jmp(L_checkRangeAndReturn);
```

Comment:
Why do we have to checkRange here, would it not be always correct? It so we could return r11 directly (by moving into rax).

Reply:
There are cases where r11 could have a value that, when added to (k - 1) would go past the end of the haystack.  I did all in my power to ensure that it doesn't but there's no test I know of to ensure that condition.  I would recommend leaving this in for now.

Reply:
Removed checkRangeAndReturn